File: atcoder/indeednow-finala-openD.py
import heapq

n=int(input())
g=[[] for _ in range(n+1)]
dp=[[[-1,-1],[-1,-1]] for _ in range(n+1)]
nv=[[] for _ in range(n+1)]
for i in range(1,n):
  a,b,t=map(int,input().split())
  if t==1:
    g[a].append((a,b,1))
    g[b].append((b,a,2))
  else:
    g[a].append((a,b,3))
    g[b].append((b,a,3))

def rec(v,p):
  nv[v]=[0,0]
  dp[v][0][0]=dp[v][1][0]=0
  for f,t,e in g[v]:
    if t==p: continue
    rec(t,v)
    for k in [0,1]:
      if e&(k+1)==0: continue
      dpk=dp[v][k]
      for j in dp[t][k]:
        if j<0 or nv[v][k]==t: continue
        if dpk[0]<j+1:
          dpk[0],dpk[1]=j+1,dpk[0]
          nv[v][k]=t
        elif dpk[1]<j+1:
          dpk[1]=j+1

rec(1, -1)
print(max(sum([[a+d-1,b+c-1] if e==f else [a+d-1,b+c-1,a+c-1] for ((a,b),(c,d)),(e,f) in zip(dp[1:],nv[1:])],[])))

File: atcoder/indeednow-finala-openD.rb
n=gets.to_i
g=(0..n).map{[]}
dp=[]
nv=[]
(n-1).times{
  a,b,t=gets.split.map(&:to_i)
  if t==1
    g[a]<<[b,1]
    g[b]<<[a,2]
  else
    g[a]<<[b,3]
    g[b]<<[a,3]
  end
}
rec=->(v,p){
  nv[v]=nvv=[0,0]
  dp[v]=[[0],[0]]
  g[v].each{|t,e|
    next if t==p
    rec[t,v]
    [0,1].each{|k|
      next if e&(k+1)==0
      dpk=dp[v][k]
      dp[t][k].each{|j|
        next if nvv[k]==t
        if dpk[0]<j+1
          dpk[0],dpk[1]=j+1,dpk[0]
          nvv[k]=t
        elsif !dpk[1] || dpk[1]<j+1
          dpk[1]=j+1
        end
      }
    }
  }
}

rec[1,-1]
p dp.zip(nv)[1,n].flat_map{|((a,b),(c,d)),(e,f)|[(d)?(a+d-1):0,(b)?(b+c-1):0,(e==f)?(a+c-1):0]}.max